ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt

```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
-                if (!call.isHandled) {
-                    call.respond(UnauthorizedResponse())
-                }
-                return@on
-            }
-            if (pluginConfig.optional &&
-                context.allFailures.none { it == AuthenticationFailedCause.InvalidCredentials }
-            ) {
-                return@on
-            }
+            if (authenticationContext._principal.principals.isNotEmpty()) return@on
+            val isOptional = optionalProviders.isNotEmpty() &&
+                firstSuccessfulProviders.isEmpty() &&
+                requiredProviders.isEmpty()
+            val isNoInvalidCredentials = authenticationContext.allFailures
+                .none { it == AuthenticationFailedCause.InvalidCredentials }
+            if (isOptional && isNoInvalidCredentials) return@on
 
-            val challenges = context.challenge.challenges
+            executeChallenges(authenticationContext, call)
+        }
+    }
 
-            for (challenge in challenges) {
-                challenge(context.challenge, call)
-                if (context.challenge.completed) {
-                    if (!call.isHandled) {
-                        call.respond(UnauthorizedResponse())
-                    }
-                    return@on
-                }
+private suspend fun executeChallenges(context: AuthenticationContext, call: ApplicationCall) {
+    val challenges = context.challenge.challenges
+
+    for (challenge in challenges) {
+        challenge(context.challenge, call)
+        if (context.challenge.completed) {
+            if (!call.isHandled) {
+                call.respond(UnauthorizedResponse())
             }
+            return
+        }
+    }
 
-            for (challenge in context.challenge.errorChallenges) {
-                challenge(context.challenge, call)
-                if (context.challenge.completed) {
-                    if (!call.isHandled) {
-                        call.respond(UnauthorizedResponse())
-                    }
-                    return@on
-                }
+    for (challenge in context.challenge.errorChallenges) {
+        challenge(context.challenge, call)
+        if (context.challenge.completed) {
+            if (!call.isHandled) {
+                call.respond(UnauthorizedResponse())
             }
+            return
+        }
+    }
 
-            for (error in context.allErrors) {
-                if (!context.challenge.completed) {
-                    logger.trace("Responding unauthorized because of error ${error.message}")
-                    if (!call.isHandled) {
-                        call.respond(UnauthorizedResponse())
-                    }
-                    context.challenge.complete()
-                    return@on
-                }
+    for (error in context.allErrors) {
+        if (!context.challenge.completed) {
+            logger.trace("Responding unauthorized because of error ${error.message}")
```

Comment:
please consider adding `call.request.uri` here

Reply:
we should also log in other cases

Reply:
I'm waiting for your PR with logging to be merged to update mine after